lib/flame/code_sync.ex

```diff
@@ -0,0 +1,311 @@
+            deleted_paths: [],
+            purge_modules: [],
+            verbose: false
+
+  def new(opts \\ []) do
+    Keyword.validate!(opts, [
+      :tmp_dir,
+      :extract_dir,
+      :copy_paths,
+      :sync_beams,
+      :start_apps,
+      :verbose
+    ])
+
+    copy_paths =
+      case Keyword.fetch(opts, :copy_paths) do
+        val when val in [{:ok, false}, :error] ->
+          []
+
+        {:ok, paths} when is_list(paths) ->
+          paths
+
+        {:ok, true} ->
+          otp_lib = :code.lib_dir()
+
+          reject_apps =
+            for app <- [:flame, :eex, :elixir, :ex_unit, :iex, :logger, :mix],
+                ebin = :code.lib_dir(app, :ebin),
+                is_list(ebin),
+                do: ebin
+
+          :code.get_path()
+          |> Kernel.--([~c"." | reject_apps])
+          |> Enum.reject(fn path -> List.starts_with?(path, otp_lib) end)
+          |> Enum.map(&to_string/1)
+      end
+
+    compute_changes(%CodeSync{
+      id: System.unique_integer([:positive]),
+      copy_paths: copy_paths,
+      sync_beams: Keyword.get(opts, :sync_beams, []),
+      tmp_dir: Keyword.get(opts, :tmp_dir, &System.tmp_dir!/0),
+      extract_dir: Keyword.get(opts, :extract_dir, fn -> "/" end),
+      start_apps: Keyword.get(opts, :start_apps, true),
+      verbose: Keyword.get(opts, :verbose, false)
+    })
+  end
+
+  def compute_changes(%CodeSync{} = code) do
+    copy_files = lookup_copy_files(code.copy_paths)
+    sync_beams_files = lookup_sync_beams_files(code.sync_beams)
+    all_paths = Enum.uniq(copy_files ++ sync_beams_files)
+
+    beam_hashes =
+      for path <- sync_beams_files,
+          into: %{},
+          do: {path, :erlang.md5(File.read!(path))}
+
+    apps_to_start =
+      case code.start_apps do
+        true ->
+          Enum.map(Application.started_applications(), fn {app, _desc, _vsn} -> app end)
+
+        false ->
+          []
+
+        apps when is_list(apps) ->
+          apps
+      end
+
+    %CodeSync{
+      code
+      | changed_paths: all_paths,
+        sync_beam_hashes: beam_hashes,
+        apps_to_start: apps_to_start
+    }
+  end
+
+  def changed?(%CodeSync{} = code) do
+    code.changed_paths != [] or code.deleted_paths != [] or code.purge_modules != []
+  end
+
+  def diff(%CodeSync{sync_beam_hashes: prev_hashes} = prev) do
+    current = compute_changes(%CodeSync{prev | copy_paths: []})
+
+    changed =
+      for kv <- current.sync_beam_hashes,
+          {path, current_hash} = kv,
+          current_hash != prev_hashes[path],
+          do: path
+
+    deleted_paths =
+      for kv <- prev.sync_beam_hashes,
+          {path, _prev_hash} = kv,
+          not Map.has_key?(current.sync_beam_hashes, path),
+          do: path
+
+    module_to_purge =
+      for path <- deleted_paths,
+          basename = Path.basename(path),
+          [mod | _] = String.split(basename, ".beam"),
+          mod != basename,
+          do: Module.concat([mod])
```

Comment:
```suggestion
          do: path |> Path.basename(".beam") |> String.to_atom()
```

since:

```
iex(6)> Path.basename("foo/Elixir.Foo.beam", ".beam") |> String.to_atom
Foo
```

and deleted_paths by definition has a `.beam` in it.